crates/diesel_models/src/query/generics.rs

```diff
@@ -22,25 +22,80 @@ use diesel::{
 use error_stack::{report, IntoReport, ResultExt};
 use router_env::{instrument, logger, tracing};
 
-use crate::{errors, PgPooledConn, StorageResult};
+use crate::{
+    errors::{self},
+    PgPooledConn, StorageResult,
+};
+
+mod db_metrics {
+    use router_env::opentelemetry::KeyValue;
+
+    #[derive(Debug)]
+    pub enum DatabaseOperation {
+        FindOne,
+        Filter,
+        Update,
+        Insert,
+        Delete,
+        DeleteWithResult,
+        UpdateWithResults,
+        UpdateOne,
+    }
+
+    #[inline]
+    pub(super) async fn track_database_call<T, Fut, U>(
+        future: Fut,
+        operation: DatabaseOperation,
+    ) -> U
+    where
+        Fut: std::future::Future<Output = U>,
+    {
+        let start = std::time::Instant::now();
+        let output = future.await;
+        let time_elapsed = start.elapsed();
+
+        let table_name = std::any::type_name::<T>().rsplit("::").nth(1);
+
+        let attributes = [
+            KeyValue::new("table", table_name.unwrap_or("undefined")),
+            KeyValue::new("operation", format!("{:?}", operation)),
+        ];
+
+        crate::metrics::DATABASE_CALLS_COUNT.add(&crate::metrics::CONTEXT, 1, &attributes);
+        crate::metrics::DATABASE_CALL_TIME.record(
+            &crate::metrics::CONTEXT,
+            time_elapsed.as_secs_f64(),
+            &attributes,
+        );
+
+        output
+    }
+}
+
+use db_metrics::*;
 
 #[instrument(level = "DEBUG", skip_all)]
 pub async fn generic_insert<T, V, R>(conn: &PgPooledConn, values: V) -> StorageResult<R>
 where
-    T: HasTable<Table = T> + Table + 'static,
+    T: HasTable<Table = T> + Table + 'static + Debug,
     V: Debug + Insertable<T>,
-    <T as QuerySource>::FromClause: QueryFragment<Pg>,
+    <T as QuerySource>::FromClause: QueryFragment<Pg> + Debug,
     <V as Insertable<T>>::Values: CanInsertInSingleQuery<Pg> + QueryFragment<Pg> + 'static,
     InsertStatement<T, <V as Insertable<T>>::Values>:
         AsQuery + LoadQuery<'static, PgConnection, R> + Send,
     R: Send + 'static,
 {
     let debug_values = format!("{values:?}");
+    let table_name = std::any::type_name::<T>().rsplit("::").nth(1);
+    logger::error!("YEE: {:?}", table_name);
```

Comment:
`YEE` 🙂

Reply:
I'll change it to a more informative message.

Reply:
Also, is there any better way to get table name, type name though simple i feel will incur some toll, does `Table` has something that we can use. Any thoughts?